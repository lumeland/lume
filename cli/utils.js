import { existsSync } from "../deps/fs.ts";
import lume, { overrides } from "../mod.js";
import { join, resolve, toFileUrl } from "../deps/path.ts";
import { bold, brightGreen, dim, red } from "../deps/colors.ts";

/**
 * Create and configure a site instance
 *
 * @param {Object} options user options to override
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  overrides.cwd = options.root;

  if (options.dev) {
    overrides.dev = options.dev;
  }

  if (options.location) {
    overrides.location = new URL(options.location);
  }

  if (options.src) {
    overrides.src = options.src;
  }

  if (options.dest) {
    overrides.dest = options.dest;
  }

  if (options.port) {
    (overrides.server ||= {}).port = parseInt(options.port);
  }

  if (options.open) {
    (overrides.server ||= {}).open = options.open;
  }

  if (options["--"]) {
    overrides.flags = options["--"];
  }

  if (existsSync(configFile)) {
    const mod = await import(toFileUrl(configFile));
    return mod.default;
  }

  return lume();
}

/**
 * Validate the number of arguments passed to a command
 *
 * @param {string} command name of the command you're validating
 * @param {[]} args array of cli arguments (i.e. options._)
 * @param {number} max the max number of args
 * @param {number} min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}

export function getCurrentVersion() {
  const url = new URL("../", import.meta.url).pathname;
  return url.match(/@([^/]+)/)?.[1] ?? `local (${url})`;
}

export function printError(exception, debug = false) {
  console.log();
  console.error(bold(red(`Error:`)), exception.message);

  printDataError(exception);

  if (debug) {
    console.error(exception);
  }

  if (exception.error) {
    let { error } = exception;
    let indent = "  ";
    while (error) {
      console.log();
      console.log(red(`${indent}${error.message}`));
      printDataError(error, indent);
      indent += "  ";

      if (debug) {
        console.error(error);
      }

      error = error.error;
    }
  }

  console.log();
}

function printDataError(error, indent = "") {
  if (!error.data) {
    return;
  }

  for (let [key, value] of Object.entries(error.data)) {
    if (key === "page") {
      value = value.src.path + value.src.ext;
    }

    console.log(dim(`${indent}${key}:`), value);
  }
}
