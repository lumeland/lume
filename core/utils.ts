import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import { SEP } from "../deps/path.ts";

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

/**
 * The list of supported MIME types.
 * It's used by the server and some plugins.
 */
export const mimes: Map<string, string> = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type>,
) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

/** Check if the argument passed is a plain object */
function isPlainObject(obj: unknown) {
  return typeof obj === "object" && obj !== null &&
    obj.toString() === "[object Object]";
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/").
 */
export function normalizePath(path: string) {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

/**
 * Search an extension in a map.
 * It's useful for cases in which the extension is multiple.
 * Example: page.tmpl.js
 */
export function searchByExtension<Type>(
  path: string,
  extensions: Map<string, Type>,
): [string, Type] | undefined {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return documentElement?.outerHTML || "";
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/**
 * Generic Exception to throw errors.
 * It allows to include extra data and the previous exception.
 */
export class Exception extends Error {
  data?: Record<string, unknown>;
  error?: Error;

  constructor(message: string, data?: Record<string, unknown>, error?: Error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
