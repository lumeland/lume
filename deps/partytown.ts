export {
  type PartytownConfig,
  partytownSnippet,
} from "npm:@qwik.dev/partytown@0.11.2/integration";
export const lib = "npm:@qwik.dev/partytown@0.11.2/lib";
