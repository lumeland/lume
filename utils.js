import { DOMParser } from "./deps/dom.ts";
import { SEP } from "./deps/path.ts";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

function isPlainObject(obj) {
  return typeof obj === "object" && obj.toString() === "[object Object]";
}

export function normalizePath(path) {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

export function documentToString(document) {
  const { doctype } = document;

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${document.documentElement.outerHTML}`;
}

const parser = new DOMParser();

export function stringToDocument(string) {
  return parser.parseFromString(string, "text/html");
}

export class Exception extends Error {
  constructor(message, data, error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
