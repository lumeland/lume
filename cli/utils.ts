import { Page, Site } from "../core.ts";
import lume from "../mod.ts";
import { exists } from "../deps/fs.ts";
import { join, resolve, toFileUrl } from "../deps/path.ts";
import { bold, dim, red } from "../deps/colors.ts";
import { Exception } from "../core/utils.ts";

/** Return the current installed version */
export function getCurrentVersion(): string {
  const url = new URL("../", import.meta.url).pathname;
  return url.match(/@([^/]+)/)?.[1] ?? `local (${url})`;
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLastVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLastDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  root = resolve(Deno.cwd(), root);

  if (config) {
    const path = join(root, config);

    if (await exists(path)) {
      console.log(`Loading config file ${dim(config)}`);
      console.log();
      const mod = await import(toFileUrl(path).href);
      return mod.default;
    }

    throw new Exception("Config file not found", { path });
  }

  const files = ["_config.js", "_config.ts"];

  for (const file of files) {
    const path = join(root, file);

    if (await exists(path)) {
      console.log(`Loading config file ${dim(file)}`);
      console.log();
      const mod = await import(toFileUrl(path).href);
      return mod.default;
    }
  }

  return lume();
}

/** Pretty-print an Error or Exception instance */
export function printError(
  error: Error | Exception,
  indent = 0,
  stackLines = 1,
) {
  console.log();
  const tab = "  ".repeat(indent);

  console.error(`${tab}${bold(red(`Error:`))}`, error.message);

  if (error instanceof Exception) {
    for (let [key, value] of Object.entries(error.data ?? {})) {
      if (key === "page") {
        value = (value as Page).src.path + (value as Page).src.ext;
      }
      console.log(dim(`${tab}${key}:`), value);
    }
  }

  if (error.stack) {
    const stack = error.stack.split("\n");

    // Skip all the stack lines that have been already presented
    stack.slice(1, stack.length - stackLines).forEach((line) => {
      console.log(`${tab}${line.trim()}`);
      stackLines++;
    });
  }

  if (error instanceof Exception && error.error) {
    printError(error.error, indent + 1, stackLines);
  }

  if (indent == 0) {
    console.log();
  }
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "bundler",
  "code_highlight",
  "date",
  "eta",
  "inline",
  "jsx",
  "postcss",
  "pug",
  "relative_urls",
  "slugify_urls",
  "svg",
  "terser",
];
