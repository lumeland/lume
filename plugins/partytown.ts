import { merge } from "../core/utils/object.ts";
import { lib, partytownSnippet } from "../deps/partytown.ts";
import type { PartytownConfig } from "../deps/partytown.ts";
import type Site from "../core/site.ts";

export interface Options {
  options?: PartytownConfig;
}

export const defaults: Options = {
  options: {
    lib: "/~partytown/",
  },
};

export function partytown(userOptions?: Options) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    const dest = options.options.lib!;
    const snippet = partytownSnippet({
      ...options.options,
      lib: site.url(dest),
    });

    const src = options.options.debug ? `${lib}/**/*.js` : `${lib}/*.js`;

    site.add(src, dest);
    site.process([".html"], (pages) => {
      for (const page of pages) {
        const script = page.document.createElement("script");
        script.textContent = snippet;
        page.document.head.appendChild(script);
      }
    });
  };
}

export default partytown;
