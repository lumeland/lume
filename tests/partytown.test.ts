import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import partytown from "../plugins/partytown.ts";

Deno.test("partytown plugin", async (t) => {
  const site = getSite({
    src: "partytown",
  });

  site.use(partytown());

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("partytown plugin (debug mode)", async (t) => {
  const site = getSite({
    src: "partytown",
  });

  site.use(partytown({
    options: {
      debug: true,
    },
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
