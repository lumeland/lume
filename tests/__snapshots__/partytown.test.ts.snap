export const snapshot = {};

snapshot[`partytown plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".html",
    },
  ],
  src: [
    "/",
    "/index.vto",
    "/~partytown",
    "/~partytown/partytown-atomics.js",
    "/~partytown/partytown-media.js",
    "/~partytown/partytown-sw.js",
    "/~partytown/partytown.js",
  ],
}
`;

snapshot[`partytown plugin 2`] = `
[
  {
    entry: "/~partytown/partytown-atomics.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown-atomics.js",
  },
  {
    entry: "/~partytown/partytown-media.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown-media.js",
  },
  {
    entry: "/~partytown/partytown-sw.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown-sw.js",
  },
  {
    entry: "/~partytown/partytown.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown.js",
  },
]
`;

snapshot[`partytown plugin 3`] = `
[
  {
    content: \`<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>title</title>
<script>!(function(w,p,f,c){if(!window.crossOriginIsolated && !navigator.serviceWorker) return;c=w[p]=Object.assign(w[p]||{},{"lib":"/~partytown/"});c[f]=(c[f]||[])})(window,'partytown','forward');/* Partytown 0.11.2 - MIT QwikDev */
const t={preserveBehavior:!1},e=e=>{if("string"==typeof e)return[e,t];const[n,r=t]=e;return[n,{...t,...r}]},n=Object.freeze((t=>{const e=new Set;let n=[];do{Object.getOwnPropertyNames(n).forEach((t=>{"function"==typeof n[t]&&e.add(t)}))}while((n=Object.getPrototypeOf(n))!==Object.prototype);return Array.from(e)})());!function(t,r,o,i,a,s,c,l,d,p,u=t,f){function h(){f||(f=1,"/"==(c=(s.lib||"/~partytown/")+(s.debug?"debug/":""))[0]&&(d=r.querySelectorAll('script[type="text/partytown"]'),i!=t?i.dispatchEvent(new CustomEvent("pt1",{detail:t})):(l=setTimeout(v,(null==s?void 0:s.fallbackTimeout)||1e4),r.addEventListener("pt0",w),a?y(1):o.serviceWorker?o.serviceWorker.register(c+(s.swPath||"partytown-sw.js"),{scope:c}).then((function(t){t.active?y():t.installing&&t.installing.addEventListener("statechange",(function(t){"activated"==t.target.state&&y()}))}),console.error):v())))}function y(e){p=r.createElement(e?"script":"iframe"),t._pttab=Date.now(),e||(p.style.display="block",p.style.width="0",p.style.height="0",p.style.border="0",p.style.visibility="hidden",p.setAttribute("aria-hidden",!0)),p.src=c+"partytown-"+(e?"atomics.js?v=0.11.2":"sandbox-sw.html?"+t._pttab),r.querySelector(s.sandboxParent||"body").appendChild(p)}function v(n,o){for(w(),i==t&&(s.forward||[]).map((function(n){const[r]=e(n);delete t[r.split(".")[0]]})),n=0;n<d.length;n++)(o=r.createElement("script")).innerHTML=d[n].innerHTML,o.nonce=s.nonce,r.head.appendChild(o);p&&p.parentNode.removeChild(p)}function w(){clearTimeout(l)}s=t.partytown||{},i==t&&(s.forward||[]).map((function(r){const[o,{preserveBehavior:i}]=e(r);u=t,o.split(".").map((function(e,r,o){var a;u=u[o[r]]=r+1<o.length?u[o[r]]||(a=o[r+1],n.includes(a)?[]:{}):(()=>{let e=null;if(i){const{methodOrProperty:n,thisObject:r}=((t,e)=>{let n=t;for(let t=0;t<e.length-1;t+=1)n=n[e[t]];return{thisObject:n,methodOrProperty:e.length>0?n[e[e.length-1]]:void 0}})(t,o);"function"==typeof n&&(e=(...t)=>n.apply(r,...t))}return function(){let n;return e&&(n=e(arguments)),(t._ptf=t._ptf||[]).push(o,arguments),n}})()}))})),"complete"==r.readyState?h():(t.addEventListener("DOMContentLoaded",h),t.addEventListener("load",h))}(window,document,navigator,top,window.crossOriginIsolated);</script></head>
<body>


</body></html>\`,
    data: {
      basename: "",
      children: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>title</title>
</head>
<body>

</body>
</html>',
      content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>title</title>
</head>
<body>

</body>
</html>',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "isCopy",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/",
    },
    src: {
      ext: ".vto",
      path: "/index",
      remote: undefined,
    },
  },
]
`;

snapshot[`partytown plugin (debug mode) 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".html",
    },
  ],
  src: [
    "/",
    "/index.vto",
    "/~partytown",
    "/~partytown/debug",
    "/~partytown/debug/partytown-atomics.js",
    "/~partytown/debug/partytown-media.js",
    "/~partytown/debug/partytown-sandbox-sw.js",
    "/~partytown/debug/partytown-sw.js",
    "/~partytown/debug/partytown-ww-atomics.js",
    "/~partytown/debug/partytown-ww-sw.js",
    "/~partytown/debug/partytown.js",
    "/~partytown/partytown-atomics.js",
    "/~partytown/partytown-media.js",
    "/~partytown/partytown-sw.js",
    "/~partytown/partytown.js",
  ],
}
`;

snapshot[`partytown plugin (debug mode) 2`] = `
[
  {
    entry: "/~partytown/debug/partytown-atomics.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown-atomics.js",
  },
  {
    entry: "/~partytown/debug/partytown-media.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown-media.js",
  },
  {
    entry: "/~partytown/debug/partytown-sandbox-sw.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown-sandbox-sw.js",
  },
  {
    entry: "/~partytown/debug/partytown-sw.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown-sw.js",
  },
  {
    entry: "/~partytown/debug/partytown-ww-atomics.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown-ww-atomics.js",
  },
  {
    entry: "/~partytown/debug/partytown-ww-sw.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown-ww-sw.js",
  },
  {
    entry: "/~partytown/debug/partytown.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/debug/partytown.js",
  },
  {
    entry: "/~partytown/partytown-atomics.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown-atomics.js",
  },
  {
    entry: "/~partytown/partytown-media.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown-media.js",
  },
  {
    entry: "/~partytown/partytown-sw.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown-sw.js",
  },
  {
    entry: "/~partytown/partytown.js",
    flags: [
      "remote",
    ],
    outputPath: "/~partytown/partytown.js",
  },
]
`;

snapshot[`partytown plugin (debug mode) 3`] = `
[
  {
    content: \`<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>title</title>
<script>!(function(w,p,f,c){if(!window.crossOriginIsolated && !navigator.serviceWorker) return;c=w[p]=Object.assign(w[p]||{},{"lib":"/~partytown/","debug":true});c[f]=(c[f]||[])})(window,'partytown','forward');/* Partytown 0.11.2 - MIT QwikDev */
const t={preserveBehavior:!1},e=e=>{if("string"==typeof e)return[e,t];const[n,r=t]=e;return[n,{...t,...r}]},n=Object.freeze((t=>{const e=new Set;let n=[];do{Object.getOwnPropertyNames(n).forEach((t=>{"function"==typeof n[t]&&e.add(t)}))}while((n=Object.getPrototypeOf(n))!==Object.prototype);return Array.from(e)})());!function(t,r,o,i,a,s,c,l,d,p,u=t,f){function h(){f||(f=1,"/"==(c=(s.lib||"/~partytown/")+(s.debug?"debug/":""))[0]&&(d=r.querySelectorAll('script[type="text/partytown"]'),i!=t?i.dispatchEvent(new CustomEvent("pt1",{detail:t})):(l=setTimeout(v,(null==s?void 0:s.fallbackTimeout)||1e4),r.addEventListener("pt0",w),a?y(1):o.serviceWorker?o.serviceWorker.register(c+(s.swPath||"partytown-sw.js"),{scope:c}).then((function(t){t.active?y():t.installing&&t.installing.addEventListener("statechange",(function(t){"activated"==t.target.state&&y()}))}),console.error):v())))}function y(e){p=r.createElement(e?"script":"iframe"),t._pttab=Date.now(),e||(p.style.display="block",p.style.width="0",p.style.height="0",p.style.border="0",p.style.visibility="hidden",p.setAttribute("aria-hidden",!0)),p.src=c+"partytown-"+(e?"atomics.js?v=0.11.2":"sandbox-sw.html?"+t._pttab),r.querySelector(s.sandboxParent||"body").appendChild(p)}function v(n,o){for(w(),i==t&&(s.forward||[]).map((function(n){const[r]=e(n);delete t[r.split(".")[0]]})),n=0;n<d.length;n++)(o=r.createElement("script")).innerHTML=d[n].innerHTML,o.nonce=s.nonce,r.head.appendChild(o);p&&p.parentNode.removeChild(p)}function w(){clearTimeout(l)}s=t.partytown||{},i==t&&(s.forward||[]).map((function(r){const[o,{preserveBehavior:i}]=e(r);u=t,o.split(".").map((function(e,r,o){var a;u=u[o[r]]=r+1<o.length?u[o[r]]||(a=o[r+1],n.includes(a)?[]:{}):(()=>{let e=null;if(i){const{methodOrProperty:n,thisObject:r}=((t,e)=>{let n=t;for(let t=0;t<e.length-1;t+=1)n=n[e[t]];return{thisObject:n,methodOrProperty:e.length>0?n[e[e.length-1]]:void 0}})(t,o);"function"==typeof n&&(e=(...t)=>n.apply(r,...t))}return function(){let n;return e&&(n=e(arguments)),(t._ptf=t._ptf||[]).push(o,arguments),n}})()}))})),"complete"==r.readyState?h():(t.addEventListener("DOMContentLoaded",h),t.addEventListener("load",h))}(window,document,navigator,top,window.crossOriginIsolated);</script></head>
<body>


</body></html>\`,
    data: {
      basename: "",
      children: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>title</title>
</head>
<body>

</body>
</html>',
      content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>title</title>
</head>
<body>

</body>
</html>',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "isCopy",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/",
    },
    src: {
      ext: ".vto",
      path: "/index",
      remote: undefined,
    },
  },
]
`;
